spacy/pipeline/pipes.pyx

```diff
@@ -1063,51 +1064,205 @@ cdef class EntityRecognizer(Parser):
+        xp = get_array_module(yh)
+        norm_yh = xp.linalg.norm(yh, axis=1, keepdims=True)
+        norm_y = xp.linalg.norm(y, axis=1, keepdims=True)
+        mul_norms = norm_yh * norm_y
+        cos = (yh * y).sum(axis=1, keepdims=True) / mul_norms
+        d_yh = (y / mul_norms) - (cos * (yh / norm_yh ** 2))
+        loss = xp.abs(cos - t).sum()
+        inverse = np.asarray([int(t[i][0]) * d_yh[i] for i in range(len(t))])
+        return loss, -inverse
 
     def __call__(self, doc):
-        self.set_annotations([doc], scores=None, tensors=None)
+        entities, kb_ids = self.predict([doc])
+        self.set_annotations([doc], entities, kb_ids)
         return doc
 
     def pipe(self, stream, batch_size=128, n_threads=-1):
-        """Apply the pipe to a stream of documents.
-        Both __call__ and pipe should delegate to the `predict()`
-        and `set_annotations()` methods.
-        """
         for docs in util.minibatch(stream, size=batch_size):
             docs = list(docs)
-            self.set_annotations(docs, scores=None, tensors=None)
+            entities, kb_ids = self.predict(docs)
+            self.set_annotations(docs, entities, kb_ids)
             yield from docs
 
-    def set_annotations(self, docs, scores, tensors=None):
-        """
-        Currently implemented as taking the KB entry with highest prior probability for each named entity
-        TODO: actually use context etc
-        """
+    def predict(self, docs):
+        self.require_model()
+        self.require_kb()
+
+        final_entities = list()
+        final_kb_ids = list()
+
+        if not docs:
+            return final_entities, final_kb_ids
+
+        if isinstance(docs, Doc):
+            docs = [docs]
+
         for i, doc in enumerate(docs):
-            for ent in doc.ents:
-                candidates = self.kb.get_candidates(ent.text)
-                if candidates:
-                    best_candidate = max(candidates, key=lambda c: c.prior_prob)
-                    for token in ent:
-                        token.ent_kb_id_ = best_candidate.entity_
+            if len(doc) > 0:
+                context_encoding = self.model([doc])
+                context_enc_t = np.transpose(context_encoding)
+                for ent in doc.ents:
+                    candidates = self.kb.get_candidates(ent.text)
+                    if candidates:
+                        scores = list()
+                        for c in candidates:
+                            prior_prob = c.prior_prob * self.prior_weight
+                            kb_id = c.entity_
+                            entity_encoding = c.entity_vector
+                            sim = float(cosine(np.asarray([entity_encoding]), context_enc_t)) * self.context_weight
+                            score = prior_prob + sim - (prior_prob*sim)
+                            scores.append(score)
+
```

Comment:
Is this loop a performance problem? There are a couple of things we could do if so. Basically we can assemble arrays and then do the operations vectorised. I often find this harder to read then a good honest loop though, so I'd only want to make the change if there's actually a problem.

Reply:
Good point @honnibal. I looked into this, testing on a batch of 10K docs, the original code took about 20s. The biggest issue was actually running each doc through the model instead of just running `self.model(docs)` once (little silly of me). This shaves off 7s, and this call to the model then takes up the most time, with about 5s out of 13s left to the `candidates` loop. I tried rewriting it using arrays (cf. latest commit), then it goes to about 3s orso. I can run a test with bigger datasets, but for now this seems the best I can do - let me know if you have further comments/ideas !